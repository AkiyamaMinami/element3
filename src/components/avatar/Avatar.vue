<script setup lang="ts">
import { computed, ref, withDefaults } from 'vue'

type AvatarSize = 'large' | 'medium' | 'small'
type AvatarFitType = 'fill' | 'contain' | 'cover' | 'none' | 'scale-down'
type AvatarShapeType = 'circle' | 'square'

interface AvatarProps {
  icon?: string
  size?: Number | AvatarSize
  shape?: AvatarShapeType
  src?: string
  srcset?: string
  alt?: string
  fit?: AvatarFitType
}

const props = withDefaults(defineProps<AvatarProps>(), {
  icon: 'el-icon-user',
  size: 'medium',
  shape: 'circle',
  fit: 'contain',
})

function useAvatarSize(props: AvatarProps) {
  return computed(() => {
    const size = Number(props.size)
    if (Number.isNaN(size))
      return
    return `
            height: ${size}px;
            width: ${size}px;
            font-size: ${size / 2.6}px;
            line-height: ${size}px;
           `
  })
}

function useClasses(props: AvatarProps) {
  return computed(() => {
    let avatarSize = ''
    if (typeof props.size === 'string')
      avatarSize = `el-avatar--${props.size}`

    return [
      'el-avatar',
      `el-avatar--${props.shape}`,
      avatarSize,
    ]
  })
}

function useFitStyle(props: AvatarProps) {
  return computed(() => {
    return `object-fit: ${props.fit}`
  })
}

const toggle = ref(true)

const avatarSize = useAvatarSize(props)
const classes = useClasses(props)
const fitStyle = useFitStyle(props)

function imgLoaded() { toggle.value = false }
</script>

<template>
  <div :class="classes" :style="avatarSize">
    <span v-if="(src || srcset)">
      <img v-show="!toggle" :src="src" :srcset="srcset" :alt="alt" :style="fitStyle" @load="imgLoaded">
    </span>
    <span v-else-if="$slots.default">
      <slot />
    </span>
    <i v-show="toggle" :class="icon" />
  </div>
</template>

<style lang="scss">
@import '../../theme/src/avatar.scss';
</style>
