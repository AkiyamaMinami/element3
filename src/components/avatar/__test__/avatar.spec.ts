import { mount } from '@vue/test-utils'
import Avatar from '../Avatar.vue'
// import { ElIcon } from '../../icon'

describe('Avatar.vue', () => {
  it('snapshot', () => {
    const wrapper = mount(Avatar)
    expect(wrapper.find('.el-avatar').exists()).toBe(true)
  })

  it('text', () => {
    const vm = mount(Avatar, {
      slots: {
        default: () => 'Im perfect avatar!',
      },
    })
    expect(vm.text()).toBe('Im perfect avatar!')
  })

  it('size', () => {
    const vm = mount(Avatar, {
      props: { size: 'large' },
    })
    expect(vm.classes()).toContain('el-avatar--large')
  })

  it('shape', () => {
    const vm = mount(Avatar, {
      props: { shape: 'square' },
    })
    expect(vm.classes()).toContain('el-avatar--square')
  })

  it('fit', () => {
    const url = 'https://cube.elemecdn.com/0/88/03b0d39583f48206768a7534e55bcpng.png'
    const vm = mount(Avatar, {
      props: { src: url, fit: 'fill' },
    })
    expect(vm.find('img').attributes('style')).toContain('object-fit: fill;')
  })

  it('icon', () => {
    const vm = mount(Avatar)
    expect(vm.find('i').classes()).toContain('el-icon-user')
    expect(vm.find('i').exists()).toBe(true)
  })
})
