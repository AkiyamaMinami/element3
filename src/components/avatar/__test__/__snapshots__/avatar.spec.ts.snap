// Vitest Snapshot v1

exports[`Avatar.vue > snapshot 1`] = `
<span
  class="el-avatar el-avatar--circle el-avatar--medium"
  style=""
>

</span>
`;
